### PR TITLE
SPEC-1045 Prohibit SCRAM negotiation and auth in SDAM

### PR DESCRIPTION
**Pull request by xdg**

### CODE REVIEW

**Comment by xdg** [comment]
The auth spec is pretty confusingly worded.  Just deleting references seemed to make it worse, so I rewrote some sections instead.  PTAL.

File: source/auth/auth.rst [hunk 1]
```diff
@@ -105,29 +105,64 @@ Naming of mechanism properties MUST be case-insensitive. For instance, SERVICE_N
...
-Connection Handshake
-~~~~~~~~~~~~~~~~~~~~
+* General-use socket: for multi-threaded drivers, these are sockets in
+  connection pools used for (non-monitoring) user operations; in
+  single-threaded drivers, these are used for both monitoring and user
+  operations.
 
-Drivers MUST consider a server ``Unknown`` if authentication fails. Effectively, an authentication failure is equivalent to a network or socket error in that we have failed to establish a connection with the server. The steps to support this are below:
+Authentication (including mechanism negotiation) MUST NOT happen on
+monitoring-only socket.
```
Comment (ajdavis) [inline source/auth/auth.rst]: "sockets"

File: source/auth/auth.rst [hunk 2]
```diff
@@ -764,15 +800,12 @@ The Java and .NET drivers currently uses eager authentication and abide by this
...
-Q: According to `Connection Handshake`_, we are calling isMaster for every socket. Isn't this a lot?
+Q: According to `Authentication Handshake`_, we are calling isMaster for every socket. Isn't this a lot?
 	Drivers should be pooling connections and, as such, new sockets getting opened should be relatively infrequent. It's simply part of the protocol for setting up a socket to be used.
 
 Q: Where is information related to user management?
 	Not here currently. Should it be? This is about authentication, not user management. Perhaps a new spec is necessary.
 
-Q: I've heard ``isMaster`` will require authentication in the future. Should we consider that here?
-	Not right now. We don't know what the future looks like yet and, as such, any preparation would be a guess. This spec will be augmented when the server changes connection protocols.
-
```
Comment (ajdavis) [inline source/auth/auth.rst]: Finally laying this to rest. =)

**Comment by xdg** [comment]
@ajdavis there's one more change based on a chat I had with Jeff.  Two parts: (a) reiterate that ismaster errors are now auth errors; (b) auth errors clear the pool.

Part (b) used to be implied (but not stated) when we said to treat auth errors as server type Unknown -- for which any driver would clear the pool.  Now that we don't say that, we should make it explicit that auth errors invalidate other sockets to the same server.

File: source/auth/auth.rst [hunk 3]
```diff
@@ -142,7 +142,9 @@ handshake:
 #. Upon opening a general-use socket to a server for a given
    MongoClient, drivers MUST issue a `MongoDB Handshake
    <../mongodb-handshake/handshake.rst>`_ immediately.  This allows a
-   driver to determine the server type.
+   driver to determine the server type.  If the ``isMaster`` of the
+   MongoDB Handshake fails with an error, drivers MUST treat this an
+   an authentication error.
```
Comment (ajdavis) [inline source/auth/auth.rst]: Meaning specifically that the server returns ok: 0? Not a network error.

File: source/auth/auth.rst [hunk 4]
```diff
@@ -155,6 +157,9 @@ handshake:
     #. A single invalid credential is the same as all credentials being
            invalid.
 
+If the authentication handshake fails for a socket, drivers MUST close all
+other general-use sockets connected to the same server.
```
Comment (ajdavis) [inline source/auth/auth.rst]: Again, an ok: 0 response to "isMaster"?

Comment (xdg) [reply source/auth/auth.rst]: Yes.  Trying to do mechanism negotiation for a non-existent user will result in an `ok:0` isMaster response.

**Comment by xdg** [comment]
Rebased and merged

**Pull request closed without merging**